File: backend/app/modules/mapping_engine/stages.py
"""Mapping stages 1, 2 and 4 as independent, individually testable scorers.

Each stage can be disabled from the engine so its contribution is measurable.
"""
from __future__ import annotations

from dataclasses import dataclass

from app.modules.mapping_engine.similarity import semantic_score
from app.modules.question_pipeline.labels import sort_key
from app.schemas.pipeline import ExtractedAnswer, ExtractedQuestion

LABEL_EXACT = 1.0
LABEL_PARENT = 0.55


@dataclass(frozen=True)
class StageWeights:
    label: float = 0.55
    spatial: float = 0.20  # U4 — reduced from 0.25 to give semantic more influence
    semantic: float = 0.25  # U4 — raised from 0.20; vision-corrected text is cleaner


DEFAULT_WEIGHTS = StageWeights()


# --------------------------------------------------------------- stage 1: explicit label
def label_score(question: ExtractedQuestion, answer: ExtractedAnswer) -> float:
    """Where a student wrote `11(a)`, that is near-certain evidence and costs nothing."""
    if not answer.detected_label:
        return 0.0
    if answer.detected_label == question.normalized_number:
        return LABEL_EXACT
    # A student writing only "11" against question "11(a)" is weak but real evidence.
    if question.normalized_number.startswith(f"{answer.detected_label}."):
        return LABEL_PARENT
    if answer.detected_label.startswith(f"{question.normalized_number}."):
        return LABEL_PARENT
    return 0.0


# U2 — Question-number offset resolver.
# Detects the integer offset between answer-sheet numbering and question-paper numbering
# from the first confident direct match, then applies it to all remaining label comparisons.
def _extract_top_int(label: str) -> int | None:
    """Return the leading integer from a normalised label, e.g. '18.a' -> 18."""
    part = label.split(".")[0]
    return int(part) if part.isdigit() else None


def detect_label_offset(
    questions: list[ExtractedQuestion],
    answers: list[ExtractedAnswer],
) -> int:
    """Scan all (question, answer) pairs for exact label matches and derive the
    most common integer offset between answer-sheet numbers and question numbers.
    Returns 0 when no offset is detectable (i.e. numbering already matches).
    """
    offsets: dict[int, int] = {}
    for answer in answers:
        if not answer.detected_label:
            continue
        a_top = _extract_top_int(answer.detected_label)
        if a_top is None:
            continue
        for question in questions:
            q_top = _extract_top_int(question.normalized_number)
            if q_top is None:
                continue
            if answer.detected_label == question.normalized_number:
                # Already matches — offset is 0 for this pair.
                offsets[0] = offsets.get(0, 0) + 1
            elif a_top - q_top != 0:
                # Candidate offset: answer number minus question number.
                diff = a_top - q_top
                offsets[diff] = offsets.get(diff, 0) + 1
    if not offsets:
        return 0
    # Return the most frequently observed offset.
    return max(offsets, key=lambda k: offsets[k])


def label_score_with_offset(
    question: ExtractedQuestion, answer: ExtractedAnswer, offset: int
) -> float:
    """label_score that also tries the offset-adjusted label before giving up."""
    base = label_score(question, answer)
    if base > 0.0 or offset == 0 or not answer.detected_label:
        return base
    a_top = _extract_top_int(answer.detected_label)
    q_top = _extract_top_int(question.normalized_number)
    if a_top is None or q_top is None:
        return 0.0
    if a_top - q_top != offset:
        return 0.0
    # Sub-part must also match when present.
    a_parts = answer.detected_label.split(".")
    q_parts = question.normalized_number.split(".")
    if a_parts[1:] == q_parts[1:]:
        return LABEL_EXACT
    if not a_parts[1:] and q_parts[1:]:
        return LABEL_PARENT
    return 0.0


# --------------------------------------------------------- stage 2: spatial / page prior
def spatial_score(
    question: ExtractedQuestion,
    answer: ExtractedAnswer,
    question_rank: dict[str, float],
    answer_rank: dict[str, float],
) -> float:
    """Most sheets are near-ordered, so preserving monotonic order is informative and free.

    This is what rescues unlabelled answers: position beats semantics on short handwriting.
    """
    q = question_rank.get(question.question_id)
    a = answer_rank.get(answer.answer_id)
    if q is None or a is None:
        return 0.0
    return round(max(0.0, 1.0 - abs(q - a)), 4)


def build_ranks(items: list[str]) -> dict[str, float]:
    """Map an ordered list of ids onto evenly spaced positions in [0, 1]."""
    if not items:
        return {}
    if len(items) == 1:
        return {items[0]: 0.5}
    return {item: index / (len(items) - 1) for index, item in enumerate(items)}


def question_order(questions: list[ExtractedQuestion]) -> list[str]:
    return [q.question_id for q in sorted(questions, key=lambda q: sort_key(q.normalized_number))]


def answer_order(answers: list[ExtractedAnswer]) -> list[str]:
    def key(answer: ExtractedAnswer) -> tuple[int, float]:
        region = answer.regions[0] if answer.regions else None
        return (answer.page_numbers[0] if answer.page_numbers else 0, region.bbox.y1 if region else 0.0)

    return [a.answer_id for a in sorted(answers, key=key)]


# ------------------------------------------------------------- stage 4: semantic overlap
def semantic_stage_score(question: ExtractedQuestion, answer: ExtractedAnswer) -> float:
    return semantic_score(question.text, answer.normalized_text)


def combine(
    label: float, spatial: float, semantic: float, weights: StageWeights = DEFAULT_WEIGHTS
) -> float:
    """An exact label is decisive on its own; otherwise the weighted blend decides."""
    if label >= LABEL_EXACT:
        return 0.97
    combined = weights.label * label + weights.spatial * spatial + weights.semantic * semantic
    # Without a label the two remaining signals must carry the full range on their own.
    if label == 0.0:
        denominator = weights.spatial + weights.semantic
        combined = (weights.spatial * spatial + weights.semantic * semantic) / denominator
        combined *= 0.85  # unlabelled matches are never as certain as labelled ones
    return round(min(0.99, max(0.0, combined)), 4)

File: apps/web/lib/server/modules/mapping_engine/stages.ts
/**
 * Mapping stages 1, 2 and 4 as independent, individually testable scorers. Port of
 * backend/app/modules/mapping_engine/stages.py. Each stage can be disabled from the
 * engine so its contribution is measurable.
 */
import { extractTopInt, sortKey, compareSortKeys } from "../question_pipeline/labels";
import { semanticScore } from "./similarity";
import type { ExtractedQuestion } from "../question_pipeline/types";
import type { ExtractedAnswer } from "../answer_pipeline/types";

export const LABEL_EXACT = 1.0;
export const LABEL_PARENT = 0.55;
// Below this, a stage's score carries no real information (rather than "mild
// disagreement") — see `combine()`.
export const WEAK_SIGNAL = 0.05;

export interface StageWeights {
  label: number;
  spatial: number;
  semantic: number;
}

export const DEFAULT_WEIGHTS: StageWeights = {
  label: 0.55,
  spatial: 0.2, // U4 — reduced from 0.25 to give semantic more influence
  semantic: 0.25, // U4 — raised from 0.20; vision-corrected text is cleaner
};

// --------------------------------------------------------------- stage 1: explicit label
/** Where a student wrote `11(a)`, that is near-certain evidence and costs nothing. */
export function labelScore(question: ExtractedQuestion, answer: ExtractedAnswer): number {
  if (!answer.detectedLabel) return 0.0;
  if (answer.detectedLabel === question.normalizedNumber) return LABEL_EXACT;
  // A student writing only "11" against question "11(a)" is weak but real evidence.
  if (question.normalizedNumber.startsWith(`${answer.detectedLabel}.`)) return LABEL_PARENT;
  if (answer.detectedLabel.startsWith(`${question.normalizedNumber}.`)) return LABEL_PARENT;
  return 0.0;
}

// U2 — Question-number offset resolver. Detects the integer offset between
// answer-sheet numbering and question-paper numbering from the first confident
// direct match, then applies it to all remaining label comparisons.
export function detectLabelOffset(questions: ExtractedQuestion[], answers: ExtractedAnswer[]): number {
  // One vote per top-level question number, not one per row: a question with several
  // sub-parts (11.a, 11.b, 11.c) must not get 3x the voting power of a question with
  // none, or a heavily sub-divided paper skews the detected offset away from the true value.
  const uniqueQTops = [
    ...new Set(
      questions
        .map((q) => extractTopInt(q.normalizedNumber))
        .filter((top): top is number => top !== null)
    ),
  ].sort((a, b) => a - b);

  const offsets = new Map<number, number>();
  for (const answer of answers) {
    if (!answer.detectedLabel) continue;
    const aTop = extractTopInt(answer.detectedLabel);
    if (aTop === null) continue;
    for (const qTop of uniqueQTops) {
      const diff = aTop - qTop;
      offsets.set(diff, (offsets.get(diff) ?? 0) + 1);
    }
  }
  if (offsets.size === 0) return 0;
  // Return the most frequently observed offset.
  let best = 0;
  let bestCount = -1;
  for (const [diff, count] of offsets) {
    if (count > bestCount) {
      best = diff;
      bestCount = count;
    }
  }
  return best;
}

/** `labelScore` that also tries the offset-adjusted label before giving up. */
export function labelScoreWithOffset(
  question: ExtractedQuestion,
  answer: ExtractedAnswer,
  offset: number
): number {
  const base = labelScore(question, answer);
  if (base > 0.0 || offset === 0 || !answer.detectedLabel) return base;
  const aTop = extractTopInt(answer.detectedLabel);
  const qTop = extractTopInt(question.normalizedNumber);
  if (aTop === null || qTop === null) return 0.0;
  if (aTop - qTop !== offset) return 0.0;
  // Sub-part must also match when present.
  const aParts = answer.detectedLabel.split(".").slice(1);
  const qParts = question.normalizedNumber.split(".").slice(1);
  if (aParts.length === qParts.length && aParts.every((p, i) => p === qParts[i])) return LABEL_EXACT;
  if (aParts.length === 0 && qParts.length > 0) return LABEL_PARENT;
  return 0.0;
}

// --------------------------------------------------------- stage 2: spatial / page prior
/**
 * Most sheets are near-ordered, so preserving monotonic order is informative and
 * free. This is what rescues unlabelled answers: position beats semantics on short
 * handwriting.
 */
export function spatialScore(
  question: ExtractedQuestion,
  answer: ExtractedAnswer,
  questionRank: Map<string, number>,
  answerRank: Map<string, number>
): number {
  const q = questionRank.get(question.questionId);
  const a = answerRank.get(answer.answerId);
  if (q === undefined || a === undefined) return 0.0;
  return Math.round(Math.max(0.0, 1.0 - Math.abs(q - a)) * 10000) / 10000;
}

/** Map an ordered list of ids onto evenly spaced positions in [0, 1]. */
export function buildRanks(items: string[]): Map<string, number> {
  const ranks = new Map<string, number>();
  if (items.length === 0) return ranks;
  if (items.length === 1) {
    ranks.set(items[0], 0.5);
    return ranks;
  }
  items.forEach((item, index) => ranks.set(item, index / (items.length - 1)));
  return ranks;
}

export function questionOrder(questions: ExtractedQuestion[]): string[] {
  return [...questions]
    .sort((a, b) => compareSortKeys(sortKey(a.normalizedNumber), sortKey(b.normalizedNumber)))
    .map((q) => q.questionId);
}

export function answerOrder(answers: ExtractedAnswer[]): string[] {
  function key(answer: ExtractedAnswer): [number, number] {
    const region = answer.regions[0];
    // An answer with no page/bbox has no real position — defaulting it to (0, 0)
    // would sort it ahead of every properly-positioned answer and shift every rank
    // after it, corrupting the spatial signal for the whole page. Push it to the end
    // instead, where it can't disturb anyone else's ordering.
    const page = answer.pageNumbers.length > 0 ? answer.pageNumbers[0] : Infinity;
    const y = region ? region.bbox.y1 : Infinity;
    return [page, y];
  }
  return [...answers]
    .sort((a, b) => {
      const [pa, ya] = key(a);
      const [pb, yb] = key(b);
      return pa !== pb ? pa - pb : ya - yb;
    })
    .map((a) => a.answerId);
}

// ------------------------------------------------------------- stage 4: semantic overlap
export function semanticStageScore(question: ExtractedQuestion, answer: ExtractedAnswer): number {
  return semanticScore(question.text, answer.normalizedText);
}

/** An exact label is decisive on its own; otherwise the weighted blend decides. */
export function combine(
  label: number,
  spatial: number,
  semantic: number,
  weights: StageWeights = DEFAULT_WEIGHTS
): number {
  if (label >= LABEL_EXACT) return 0.97;
  let combined = weights.label * label + weights.spatial * spatial + weights.semantic * semantic;
  // Without a label the two remaining signals must carry the full range on their own.
  if (label === 0.0) {
    if (spatial <= WEAK_SIGNAL && semantic > WEAK_SIGNAL) {
      // No usable positional evidence at all — an answer written out of the expected
      // order — so a strong text match must not be diluted by spatial's absent
      // weight. A plain weighted average with spatial pinned at 0 caps out around
      // 0.38-0.47 even for a perfect semantic match, below `review_threshold` (0.45):
      // legitimate out-of-order answers were silently dropped as unanswered.
      combined = semantic * 0.85;
    } else {
      const denominator = weights.spatial + weights.semantic;
      combined = (weights.spatial * spatial + weights.semantic * semantic) / denominator;
      combined *= 0.85; // unlabelled matches are never as certain as labelled ones
    }
  }
  return Math.round(Math.min(0.99, Math.max(0.0, combined)) * 10000) / 10000;
}
